music_assistant/server/providers/tidal/helpers.py

```diff
@@ -0,0 +1,493 @@
+    album = Album(item_id=album_id, provider=tidal_provider.instance_id, name=name, version=version)
+    for artist_obj in album_obj.artists:
+        album.artists.append(
+            await parse_artist(tidal_provider=tidal_provider, artist_obj=artist_obj)
+        )
+    if album_obj.type == "ALBUM":
+        album.album_type = AlbumType.ALBUM
+    elif album_obj.type == "COMPILATION":
+        album.album_type = AlbumType.COMPILATION
+    elif album_obj.type == "EP":
+        album.album_type = AlbumType.EP
+    elif album_obj.type == "SINGLE":
+        album.album_type = AlbumType.SINGLE
+
+    album.upc = album_obj.universal_product_number
+    album.year = int(album_obj.year)
+    album.add_provider_mapping(
+        ProviderMapping(
+            item_id=album_id,
+            provider_domain=tidal_provider.domain,
+            provider_instance=tidal_provider.instance_id,
+            content_type=ContentType.FLAC,
+            bit_rate=1411,
+            url=f"http://www.tidal.com/album/{album_id}",
+        )
+    )
+    album.metadata = await parse_album_metadata(tidal_provider, album_obj)
+    return album
+
+
+async def parse_album_metadata(tidal_provider, album_obj: TidalAlbum) -> MediaItemMetadata:
+    metadata = MediaItemMetadata()
+    image_url = None
+    try:
+        image_url = album_obj.image(1280)
+    except Exception:
+        tidal_provider.logger.info(f"Album {album_obj.id} has no available picture")
+    metadata.images = [
+        MediaItemImage(
+            ImageType.THUMB,
+            image_url,
+        )
+    ]
+    metadata.copyright = album_obj.copyright
+    metadata.explicit = album_obj.explicit
+    metadata.popularity = album_obj.popularity
+    return metadata
+
+
+async def parse_track(tidal_provider, track_obj: TidalTrack) -> Track:
+    """Parse tidal track object to generic layout."""
+    version = track_obj.version if track_obj.version is not None else None
+    track_id = str(track_obj.id)
+    track = Track(
+        item_id=track_id,
+        provider=tidal_provider.instance_id,
+        name=track_obj.name,
+        version=version,
+        duration=track_obj.duration,
+        disc_number=track_obj.volume_num,
+        track_number=track_obj.track_num,
+    )
+    track.isrc.add(track_obj.isrc)
+    track.album = get_item_mapping(
+        tidal_provider=tidal_provider,
+        media_type=MediaType.ALBUM,
+        key=track_obj.album.id,
+        name=track_obj.album.name,
+    )
+    track.artists = []
+    for track_artist in track_obj.artists:
+        artist = await parse_artist(tidal_provider=tidal_provider, artist_obj=track_artist)
+        track.artists.append(artist)
+    available = track_obj.available
+    track.add_provider_mapping(
+        ProviderMapping(
+            item_id=track_id,
+            provider_domain=tidal_provider.domain,
+            provider_instance=tidal_provider.instance_id,
+            content_type=ContentType.FLAC,
+            bit_rate=1411,
```

Comment:
is the contenttype flac by default ? and is the sample rate and bit depth limited to redbook (CD quality) 44.1khz/16 bits ? Tidal also has some hi res material right ? Or is that limited to MQA ?

Reply:
The higher quality stuff I thought I would add later. The plan is to give the option during setup of choosing quality. Now assuming that our token and the user account have access to that quality and the track eexists also in that quality... I would set that at parsing time

Reply:
I was giving this some thought, and I had a couple of ideas for this. However, I'll need to upgrade my account to implement them.
However, the HI RES recordings could be somewhat interesting, in so much as the bit rate, mode (stereo, Atmos, etc.) and codec) on the HI RES versions varies and may not be immediately available when parsing the track.
Bitrate varies between 2304-9216 Kbps
Not sure about codecs used yet.
Mode can be MQA, Reality360 or Atmos.

I think I can only get the codec at streaming time, although one could assume that the codec will be directly related to the audio mode.
Bit rate is not included in any response per se, however, I assume that MA will calculate the bitrate during streaming? Do I assume correctly here and If so, should I just skip setting it in the provider mapping altogether?

Reply:
bitrate is not important, bit depth and sample rate is, that should be set correctly, as well as the codec. The more accurate quality details you provide, the better. For the streamdetails these details NEED to be accurate, for just the regular track item it MAY be accurate.

Reply:
Sadly again, Tidal are rather vague about the bit depth and sample rate.
The say typically HI RES is 96KHz/24bit, but that leaves room for interpretation.
I'll dig into web console and see if I can find the info anywhere in some obscure call that isn't mapped in the tidalapi.

Reply:
Let's do that in a follow-up PR

Reply:
Yup